src/api/Api.js

```diff
@@ -95,65 +157,142 @@ export class Api {
       headers = {}
     }
 
-    this.headers = Object.assign(this.headers, headers)
+    for (const [name, value] of Object.entries(headers)) {
+      this.setHeader(name, value)
+    }
   }
 
-  async doRequest (method, url, params, data, options) {
-    const config = Object.assign(options, this.defaultOptions)
-    config.headers = this.headers
+  setBaseUrl (baseUrl) {
+    this.baseUrl = baseUrl
+  }
+
+  /* #region request methods */
+
+  _compileUrl (url, params) {
+    let urlObj = null
+    const baseUrlObj = new URL(this.baseUrl)
+
+    if (url.indexOf('//') === 0) {
+      url = baseUrlObj.protocol + url
+    }
 
     if (!isAbsoluteUri(url)) {
-      url = this.baseUrl + url
+      // various path adjustments
+      url = url.replace(/^(\/)(.+)/, '$2')
+
+      if (baseUrlObj.pathname !== '' && baseUrlObj.pathname !== '/') {
+        const deSlashedBaseUrlPath = baseUrlObj.pathname.replace(
+          /(\/)(.+)(\/)$/,
+          '$2'
+        )
+
+        url = deSlashedBaseUrlPath + '/' + url
+
+        baseUrlObj.pathname = ''
+      }
+
+      urlObj = new URL(baseUrlObj.href)
+      urlObj.pathname = url
```

Comment:
While I think the test cases for this method explain the purpose quite well, the tl;dr: of this mess is to get rid of any slashes that could be leading to double slashes when combining paths like `foo.com/api/` `/Book/1` to `foo.com/api/Book/1`.